File: kotlin-node/karakum/src/jsMain/resources/plugins/convertOverriddenPropertyDeclaration.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

const overriddenProps = {
    // net
    "Socket": {
        "destroyed": "netSocketDestroyed"
    },

    // perf_hooks
    "PerformanceMark": {
        "duration": null,
        "entryType": ["markEntryType", "EntryType"]
    },
    "PerformanceMeasure": {
        "entryType": ["measureEntryType", "EntryType"]
    },
    "PerformanceNodeTiming": {
        "entryType": ["nodeTimingEntryType", "EntryType"]
    },
    "PerformanceResourceTiming": {
        "entryType": ["resourceTimingEntryType", "EntryType"]
    },

    // stream
    "ReadableBase": {
        "readable": null
    },
    "WritableBase": {
        "writable": "readonlyWritable"
    },
    "Duplex": {
        "writable": "duplexWritable",
        "writableEnded": null,
        "writableFinished": null,
        "writableHighWaterMark": null,
        "writableLength": null,
        "writableObjectMode": null,
        "writableCorked": null,
        "writableNeedDrain": null,
        "closed": null,
        "errored": null,
    }
}

export default function (node, context, render) {
    if (
        ts.isPropertyDeclaration(node)
        && ts.isClassDeclaration(node.parent)
        && node.parent.name
        && node.parent.name.text in overriddenProps
        && ts.isIdentifier(node.name)
        && node.name.text in overriddenProps[node.parent.name.text]
    ) {
        const readonly = node.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)

        const modifier = readonly
            ? "val "
            : "var "

        const name = render(node.name)

        const isOptional = Boolean(node.questionToken)

        const type = karakum.renderNullable(node.type, isOptional, context, render)

        const overrideConfig = overriddenProps[node.parent.name?.text][node.name.text]

        const hasAlias = overrideConfig !== null

        if (hasAlias) {
            let alias
            let parentType

            if (Array.isArray(overrideConfig)) {
                [alias, parentType] = overrideConfig
            } else {
                alias = overrideConfig
            }

            const jsName = `@JsName("${name}")`
            const aliasSignature = `${modifier}${alias}: ${type}`

            if (parentType) {
                const deprecation = `@Deprecated(message = "use ${alias}", level = DeprecationLevel.HIDDEN)`
                const signature = `override ${modifier}${name}: ${parentType}`

                return [
                    deprecation,
                    signature,
                    jsName,
                    aliasSignature,
                ].join("\n")
            } else {
                return [
                    jsName,
                    aliasSignature,
                ].join("\n")
            }
        }

        return `override ${modifier}${name}: ${type}`
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertOverriddenPropertyDeclaration.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.getParentOrNull
import node.karakum.plugins.DeclarationOverrideConfig.Alias
import node.karakum.plugins.DeclarationOverrideConfig.Overlap
import node.karakum.plugins.DeclarationOverrideConfig.Preserve
import typescript.SyntaxKind
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isIdentifier
import typescript.isPropertyDeclaration


private sealed interface DeclarationOverrideConfig {
    object Preserve : DeclarationOverrideConfig
    class Alias(val alias: String) : DeclarationOverrideConfig
    class Overlap(val alias: String, val parentType: String) : DeclarationOverrideConfig
}

private val overriddenProps = mapOf(
    // net
    "Socket" to mapOf(
        "destroyed" to Alias("netSocketDestroyed")
    ),

    // perf_hooks
    "PerformanceMark" to mapOf(
        "duration" to Preserve,
        "entryType" to Overlap("markEntryType", "EntryType")
    ),
    "PerformanceMeasure" to mapOf(
        "entryType" to Overlap("measureEntryType", "EntryType")
    ),
    "PerformanceNodeTiming" to mapOf(
        "entryType" to Overlap("nodeTimingEntryType", "EntryType")
    ),
    "PerformanceResourceTiming" to mapOf(
        "entryType" to Overlap("resourceTimingEntryType", "EntryType")
    ),

    // stream
    "Readable" to mapOf(
        "readable" to Preserve
    ),
    "Writable" to mapOf(
        "writable" to Alias("readonlyWritable")
    ),
    "Duplex" to mapOf(
        "writable" to Alias("duplexWritable"),
        "writableEnded" to Preserve,
        "writableFinished" to Preserve,
        "writableHighWaterMark" to Preserve,
        "writableLength" to Preserve,
        "writableObjectMode" to Preserve,
        "writableCorked" to Preserve,
        "writableNeedDrain" to Preserve,
        "closed" to Preserve,
        "errored" to Preserve,
    )
)

val convertOverriddenPropertyDeclaration = createPlugin { node, context, render ->
    nullable {
        ensure(isPropertyDeclaration(node))

        val classNode = ensureNotNull(node.getParentOrNull())
        ensure(isClassDeclaration(classNode))

        val className = ensureNotNull(classNode.name)
        val classOverrides = ensureNotNull(overriddenProps[className.text])

        val propertyName = node.name
        ensure(isIdentifier(propertyName))
        val propertyOverride = ensureNotNull(classOverrides[propertyName.text])

        val readonly = node.modifiers?.asArray()?.find { it.kind == SyntaxKind.ReadonlyKeyword }

        val modifier = if (readonly != null) "val " else "var "

        val name = render(node.name)

        val isOptional = node.questionToken != null

        val type = renderNullable(node.type, isOptional, context, render)

        when (propertyOverride) {
            is Overlap -> {
                val alias = propertyOverride.alias
                val jsName = "@JsName(\"$name\")"
                val aliasSignature = "${modifier}${alias}: $type"
                val deprecation = "@Deprecated(message = \"use ${alias}\", level = DeprecationLevel.HIDDEN)"
                val signature = "override ${modifier}${name}: ${propertyOverride.parentType}"

                arrayOf(
                    deprecation,
                    signature,
                    jsName,
                    aliasSignature,
                ).joinToString("\n")
            }
            is Alias -> {
                val alias = propertyOverride.alias
                val jsName = "@JsName(\"$name\")"
                val aliasSignature = "${modifier}${alias}: $type"

                arrayOf(
                    jsName,
                    aliasSignature,
                ).joinToString("\n")
            }
            Preserve -> "override ${modifier}${name}: $type"
        }
    }
}
